### PR TITLE
add horizondex on base

### PR DESCRIPTION
**Pull request by pachecospig**
# Thank you for contributing to Spellbook!

Thank you for taking the time to submit code in Spellbook. A few things to consider:
- If you are a first-time contributor, please sign the CLA by copy & pasting **_exactly_** what the bot mentions in PR comment
- Refer to [docs](#spellbook-contribution-docs) section below to answer questions
- Dune team will review submitted PRs as soon as possible

## Spellbook sub-projects

To prepare for Spellbook to scale in size, Dune team has started to pilot sub-projects:
- For a quick overview, see [the GH discussion on the topic](https://github.com/duneanalytics/spellbook/discussions/5238)
    - Please ask questions and continue the conversation there
- The first spell lineage to move: `tokens.erc20` and all upstream spells which help build it
    - **_Note_**: Please navigate to [tokens](https://github.com/duneanalytics/spellbook/tree/main/tokens) root directory to find tokens objects

## Spellbook contribution docs

The [docs](https://github.com/duneanalytics/spellbook/tree/main/docs) directory has been implemented to answer as many questions as possible. Please take the time to reference each `.md` file within this directory to understand how to efficiently contribute & why the repo is designed as it is 🪄

Example questions to be answered:
- What does each property in the [model config block](https://github.com/duneanalytics/spellbook/blob/main/docs/models/model_config_block.md) mean?
- What is the [CI test](https://github.com/duneanalytics/spellbook/blob/main/docs/ci_test/ci_test_overview.md) attached to PRs and how can I best utilize it?
- What are the Spellbook [best practices](https://github.com/duneanalytics/spellbook/blob/main/docs/general/best_practices.md)?

Please navigate through the [docs](https://github.com/duneanalytics/spellbook/tree/main/docs) directory to find as much info as you can.

**Note**: happy to take PRs to improve the docs, let us know 🤝

### CODE REVIEW

**Comment by pachecospig** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by pachecospig** [comment]
Source tables for horizondex_base.Pool_evt_Swap have names 'deltaQty0' and 'deltaQty1' instead of Uniswap_V3 'amount0' and amount1' which throws off the macro.
I just rewrote macro as an SQL query in horizondex_base_base_trades.sql.

Could also change the column names in the source table for the decoded contract, or
could also edit macro:
```
{% macro uniswap_compatible_v3_trades(
	...
	amount0_column_name = 'amount0',
	amount1_column_name = 'amount1'
	)
%}
WITH dexs AS
(
	SELECT
	...
	, CASE WHEN t.{{ amount0_column_name }} < INT256 ...
	...
)
```
But I don't think this is a very common pattern.

**Comment by Hosuke** [comment]
> Source tables for horizondex_base.Pool_evt_Swap have names 'deltaQty0' and 'deltaQty1' instead of Uniswap_V3 'amount0' and amount1' which throws off the macro. I just rewrote macro as an SQL query in horizondex_base_base_trades.sql.
> 
> Could also change the column names in the source table for the decoded contract, or could also edit macro:
> 
> ```
> {% macro uniswap_compatible_v3_trades(
> 	...
> 	amount0_column_name = 'amount0',
> 	amount1_column_name = 'amount1'
> 	)
> %}
> WITH dexs AS
> (
> 	SELECT
> 	...
> 	, CASE WHEN t.{{ amount0_column_name }} < INT256 ...
> 	...
> )
> ```
> 
> But I don't think this is a very common pattern.

You can check `/macros/kyberswap_compatible_trades.sql` first.

If the `horizondex_base.base_trades` is a single special case, we may implement it in full without using macros.

**Comment by pachecospig** [comment]
kyberswap_compatible_trades prepends the string `kyber_` to source tables:
```
FROM {{ source('kyber_' ~ blockchain, src["source_evt_swap"]) }} t
            INNER JOIN {{ source('kyber_' ~ blockchain, src["source_evt_factory"]) }} p ON t.contract_address = p.pool
```

I don't see any way to access source tables (like horizondex_base) that don't begin with 'kyber_' from that macro. Clipper and airswap compatible trades have a similar prepended string.
Balancer and dodo use an input (`project_decoded_as` or `decoded_project`) which allows more flexibility.
```
{% macro dodo_compatible_trades(
    ...
    decoded_project = 'dodo',
    ...
    )
%}

...
FROM {{ source(decoded_project ~ '_' ~ blockchain, sell_base_token_source )}} s
...
```

I guess options are:
Leave it in full
Write a new macro for this type
Submit a new pull for the kyberswap macro (and kyberswap_chain_base_trades for all chains).

Happy to do any of them as you see fit.

File: models/_sector/dex/trades/base/platforms/horizondex_base_base_trades.sql [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+(
+    SELECT
+        t.evt_block_number AS block_number
+        , t.evt_block_time AS block_time
+        , t.recipient AS taker
+        , cast(null as varbinary) as maker
+        , CASE WHEN t.deltaQty0 < INT256 '0' THEN abs(t.deltaQty0) ELSE abs(t.deltaQty1) END AS token_bought_amount_raw -- when amount0 is negative it means trader_a is buying token0 from the pool
+        , CASE WHEN t.deltaQty0 < INT256 '0' THEN abs(t.deltaQty1) ELSE abs(t.deltaQty0) END AS token_sold_amount_raw
+        , CASE WHEN t.deltaQty0 < INT256 '0' THEN f.token0 ELSE f.token1 END AS token_bought_address
+        , CASE WHEN t.deltaQty0 < INT256 '0' THEN f.token1 ELSE f.token0 END AS token_sold_address
```
Comment (Hosuke) [inline models/_sector/dex/trades/base/platforms/horizondex_base_base_trades.sql:23]: Hi @tomfutago,
this part seems to be very similar to [`kyberswap_compatible_trades`](https://github.com/Hosuke/spellbook/blob/main/macros/models/_sector/dex/kyberswap_compatible_trades.sql#L17), but with the tokens flipped.

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @pachecospig 

[Sample horizondex_base_base_trades](https://dune.com/queries/3431952)

File: models/_sector/dex/trades/base/platforms/horizondex_base_base_trades.sql (hunk 1 above)
Comment (jeff-dude) [reply models/_sector/dex/trades/base/platforms/horizondex_base_base_trades.sql:23]: @Hosuke if we have any doubt that the token bought/sold are flipped in any given new dex, we should force `dex.trades_beta` to be modified in PR and have that run. that way we see `amount_usd` populated and then the test on that field will run to ensure no values out of proportion due to incorrect assignment. when merging new dex PRs, `dex.trades_beta` runs in prod to refresh regardless, so modifying to run is no big deal

Comment (jeff-dude) [reply models/_sector/dex/trades/base/platforms/horizondex_base_base_trades.sql:23]: lets go ahead and try that here before merge

File: seeds/_sector/dex/horizondex_base_base_trades_seed.csv [hunk 2]
```diff
@@ -0,0 +1,4 @@
+﻿blockchain,project,version,block_date,tx_hash,evt_index,token_bought_address,token_sold_address,block_number,token_bought_amount_raw,token_sold_amount_raw
+base,horizondex,1,2023-8-7,0x093685041bc3c5461cb80f33bda28de44ace327f4881e614f3e80b0ed260b75e ,51,0x27D2DECb4bFC9C76F0309b8E88dec3a601Fe25a8,0x4200000000000000000000000000000000000006,2300086,7524272132295450150587,25500000000000010
```
[the remark this answers is not in the document]
Comment (Hosuke) [reply seeds/_sector/dex/horizondex_base_base_trades_seed.csv:2]: https://dune.com/queries/3434484

Looks lgtm.

**Review by Hosuke — APPROVED** [review]
I just ran the dex.trades_beta:
https://dune.com/queries/3434498

the amount_usd matches the value for the trade.

File: seeds/_sector/dex/horizondex_base_base_trades_seed.csv (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeff-dude) [reply seeds/_sector/dex/horizondex_base_base_trades_seed.csv:2]: thanks for taking the extra steps!

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**